test/pubsub.spec.js

```diff
@@ -6,386 +6,194 @@ chai.use(require('dirty-chai'))
 chai.use(require('chai-spies'))
 const expect = chai.expect
 const sinon = require('sinon')
-const series = require('async/series')
-const parallel = require('async/parallel')
+const DuplexPair = require('it-pair/duplex')
 
 const PubsubBaseProtocol = require('../src')
 const { randomSeqno } = require('../src/utils')
-const utils = require('./utils')
-const createNode = utils.createNode
+const { createPeerInfo, mockRegistrar, PubsubImplementation } = require('./utils')
 
-class PubsubImplementation extends PubsubBaseProtocol {
-  constructor (libp2p) {
-    super('libp2p:pubsub', 'libp2p:pubsub-implementation', libp2p)
-  }
+describe('pubsub base protocol', () => {
+  describe('should start and stop properly', () => {
+    let pubsub
+    let sinonMockRegistrar
+
+    beforeEach(async () => {
+      const peerInfo = await createPeerInfo()
+      sinonMockRegistrar = {
+        register: sinon.stub(),
+        unregister: sinon.stub()
+      }
 
-  publish (topics, messages) {
-    // ...
-  }
+      pubsub = new PubsubBaseProtocol('pubsub', '/pubsub/1.0.0', peerInfo, sinonMockRegistrar)
 
-  subscribe (topics) {
-    // ...
-  }
+      expect(pubsub.peers.size).to.be.eql(0)
+    })
 
-  unsubscribe (topics) {
-    // ...
-  }
+    afterEach(() => {
+      sinon.restore()
+    })
 
-  _processConnection (idB58Str, conn, peer) {
-    // ...
-  }
-}
+    it('should be able to start and stop', async () => {
+      await pubsub.start()
+      expect(sinonMockRegistrar.register.calledOnce).to.be.true()
 
-describe('pubsub base protocol', () => {
-  afterEach(() => {
-    sinon.restore()
-  })
-
-  describe('fresh nodes', () => {
-    let nodeA
-    let nodeB
-    let psA
-    let psB
-
-    before((done) => {
-      series([
-        (cb) => createNode(cb),
-        (cb) => createNode(cb)
-      ], (err, nodes) => {
-        if (err) {
-          return done(err)
-        }
-        nodeA = nodes[0]
-        nodeB = nodes[1]
-        done()
-      })
+      await pubsub.stop()
+      expect(sinonMockRegistrar.unregister.calledOnce).to.be.true()
     })
 
-    before('mount the pubsub protocol', (done) => {
-      psA = new PubsubImplementation(nodeA)
-      psB = new PubsubImplementation(nodeB)
+    it('should not throw to start if already started', async () => {
+      await pubsub.start()
+      await pubsub.start()
+      expect(sinonMockRegistrar.register.calledOnce).to.be.true()
 
-      setTimeout(() => {
-        expect(psA.peers.size).to.be.eql(0)
-        expect(psB.peers.size).to.be.eql(0)
-        done()
-      }, 50)
+      await pubsub.stop()
+      expect(sinonMockRegistrar.unregister.calledOnce).to.be.true()
     })
 
-    before('start both Pubsub', (done) => {
-      parallel([
-        (cb) => psA.start(cb),
-        (cb) => psB.start(cb)
-      ], done)
+    it('should throw if stop before start', async () => {
```

Comment:
I would change this behavior to not throw, it should just return and do nothing. There's no reason to throw, similar to starting when already started.